### ISSUE TITLE
promote_typejoin with abstract types

### ISSUE THREAD

**Issue by nalimilan** [JuliaLang/julia#35504]
Currently, `promote_typejoin(Missing, T)` returns `Any` when `T` isn't concrete. This can be annoying in cases where the type information is necessary to perform some operation (and not just for performance), like computing `zero(T)`:
```julia
julia> nt1 = NamedTuple{(:x, :y),Tuple{Union{Missing, Int},Union{Missing, Int}}}((missing, missing));

julia> eltype(nt1)
Union{Missing, Int64}

julia> sum(skipmissing(nt1))
0

julia> nt2 = NamedTuple{(:x, :y),Tuple{Union{Missing, Int},Union{Missing, Float64}}}((missing, missing));

julia> eltype(nt2)
Any

julia> sum(skipmissing(nt2))
ERROR: MethodError: no method matching zero(::Type{Any})
[...]
```
This would work if `promote_typejoin` returned `Union{Missing, Real}` since `zero` is defined for that type.

This situation happens in practice, for example when a row or group of rows happens to contain only missing values (see https://github.com/JuliaData/DataFrames.jl/pull/2196).

Would it make sense to change `promote_typejoin` to stop special-casing concrete types?

Cc: @JeffBezanson @vtjnash

**Comment by pdeffebach** [comment]
> This would work if `promote_type` returned `Union{Missing, Real}` since `zero` is defined for that type.

I think you mean `promote_typejoin` rather than `promote_type`. 

```
julia> promote_type(Union{Missing, Int}, Float64)
Union{Missing, Float64}
```

I don't understand the distinction, though.

**Comment by nalimilan** [comment]
Woops, that was a typo. I was indeed talking about `promote_typejoin` everywhere.

(`promote_type` is very different as it's allowed to return a type which isn't a supertype of both inputs. It's used notably by `vcat` or by fallback methods for mathematical operators e.g. to return `Float64` when the inputs mix integers and `Float64`.)

**Comment by pdeffebach** [comment]
The solution seems to be altering this block [here](https://github.com/JuliaLang/julia/blob/ba4f858cd0618e209b81a1800005662769cdc1c1/base/promotion.jl#L130) to add a method like

```
_promote_typejoin(a::Type{S}, b::Type{Union{T, Missing}}) = Union{typejoin(S, T), MIssing}
```

**Comment by nalimilan** [comment]
The `isconcretetype(T)` check can simply be removed. That's mostly a design decision.

**Comment by pdeffebach** [comment]
But we aren't joining to `Missing`, right? Those methods are just for joining `Missing` with something. Here, we are joinning `Union{T, MIssing}` with something that is *not* missing.

**Comment by pdeffebach** [comment]
Would it be possible to get something that fixes this into 1.6?

**Comment by nalimilan** [comment]
I've filed https://github.com/JuliaLang/julia/pull/36939 to get the ball rolling.

**Comment by nalimilan** [comment]
This just got merged into Julia master, so it should be in 1.6.

**Comment by pdeffebach** [comment]
Wonderful! Now basically all `egen` behaviors are easy in Julia!
